### ISSUE TITLE
Check that last tested version exists in NVDA

### ISSUE THREAD

**Issue by nvdaes** [nvaccess/addon-datastore-validation#16]
In addonFiles repo,manual reviews need to ensure that last tested version exist in NVDA:
- When a beta is published, the corresponding last tested version is considered valid, but versions cannot be greater than it, for example, a version like 9999.1.0 would be invalid, or even 2024.1.0. This is a way to skipping the compatibility procedure set in NVDA.
- Seems that this is not checked, so that we may release add-ons with a big version to avoid re-tests and updates when NVDA breaks compatibility.

A maxApiVersion may be provided in this repo, or for better maintainment, addon-datastore-transform maybe checked out in GitHub Actions to see the last version existing in NVDAApiVersions.

cc: @seanbudd

**Comment by nvdaes** [comment]
Anyway, in general I think that this should be checked even for minimum NVDA version. I think that pull request check should fail if minimum or last tested version of an add-on is not in NVDAApi versions. I've mentioned explicitly to Sean since this maybe done before enabling automerge in addon-datastore, that, according of documentation read by me, maybe to add people as a collaborator granting write permission to the repo, configuring a rule to protect master branch so that pull requests are required and writing directly there is not allowed.

**Comment by nvdaes** [comment]
I see your reaction @seanbudd . I may try to create an additional check in GitHub Actions, and maybe a pull request here, checking out the addon-datastore-transform and trying to see if minimum and last tested version exist.

**Comment by nvdaes** [comment]
I think it's not possible to check properly ninimum required NVDA versions, since they maybe even quite old like 2018.1, 2017.3, etc., and these aren't specifically listed on NVDAApiVersions. But at lest we may check if last tested version is major of the lastApiVersion, to avoid skipping the compatibility mechanism related to NVDA stability. I'm creating a branch in addon-datastore adding two extra steps to the validationMetadata workflow, to check if last tested version is greater than last API Version.

**Comment by seanbudd** [comment]
I would suggest editing validate.py to create these rules.
I think you can test minimum required versions. NVDAAPIVersions.json should contain every version since the versioning system was introduced. So any valid submission should use a valid api version for both the minimum and last tested. NVDA 2019.2 and earlier should use `0.0.0`.

**Comment by nvdaes** [comment]
OK Sean,then we need to download the NVApiVersions.json and then use rules. This may take some time. While this is not added, submissions maybe checked manually to avoid skipping re-tests and updates.

**Comment by seanbudd** [comment]
I think you should be able to pass in `NVApiVersions.json` by filename from the command line.
You can use the file in the `addon-datastore` repository actions.

**Comment by nvdaes** [comment]
@seanbudd I'm working on PR #17 for this issue. Feel free to work yourself if you think that my work is too slow or if you prefer to use a different code or whatever.
I'll be happy to work on this, but I'll do it in spare time, so don't worry if you want to supersed my PR.

**Comment by CyrilleB79** [comment]
Not sure to understand well.

Do you plan to allow both minimum NVDA version and last tested version only be valid if in the API version, i.e. 0.0.0, 2019.3, 2021.1, 2022.1 and 2023.1?

Regarding  min version, this does not seem a good design to me. Indeed, I have personally Character Information compatible only with 2022.3.3 onwards since it uses functions that have been introduced in 2022.3.3. Also Joseph has regularly modernized his add-on using newly introduced functions so that min version do not always match an API-version.

**Comment by nvdaes** [comment]
Hi Cyrille, API versions aren't versions that break changes. And I think that if some people introduces changes available in some versions, these particular versions should be added to API versions so it's clear that they exist. Otherwise, why the NVDAApiVersion file is created? Comments are appreciated.

**Comment by XLTechie** [comment]
@nvdaes, @seanbudd, just because NVDA didn't use specific API versions for many 
years, does not mean that everything written, for example, to work in NVDA 
2018.1, will work in NVDA 2015.1. Many things changed in NVDA during that time.

So if an add-on author knows that his add-on only starts working because of an 
NVDA feature introduced in 2017.3, and he wants to signal to users that this 
add-on is only tested with 2017.3 and above, how is he to do that?

Currently he would set the "addon_minimumNVDAVersion" in the manifest/template 
buildVars. Even though NVDA didn't enforce API versions on add-ons back during 
that version of NVDA, it is the only mechanism given to indicate to users what 
the add-on author expects them to understand.

Further, at least in buildVars, it is called the minimum NVDA version, not the 
minimum API version of accepted API versions. That may have always been what it 
meant from NVDA's point of view, but I think the way add-on developers looked at 
it has traditionally been: this is the lowest version of NVDA with which my 
add-on is guaranteed to work.
It's even right there in the name of the manifest parameter.

So as I understand it, and maybe I'm wrong, but: developers have traditionally 
told users the lowest version of NVDA in which the add-on works, but now will be 
forced to remove that information and use 0.0.0. By doing that, this makes them 
effectively claim that it works with any NVDA version between 0.0.0 and 2019.2.1 
(i.e. Python 2), even if they know that's incorrect.

It is not realistic to suggest that we go back, find every add-on that relies on 
a feature that was only introduced in a specific NVDA version before core came 
up with the API version idea, and add it to the API versions file. Or even to 
take every known add-on's minimum NVDA version, and add that to the API versions 
file.

Surely it would be better not to vet the minimum NVDA version if it falls below 
2019.3.1, and just accept what the author has listed?

**Comment by XLTechie** [comment]
The only way I can think of making this doable, is to go back and add to the API 
versions file, every individual version of NVDA that was released, prior to 
2019.3.1.

But if you're going to do that, why even check in the first place? What is 
gained?

Were it up to me, I would just check the format of the minimumNVDAVersion string 
to be sure it matched

^(\d{4,4}(\.\d){1,2}|0(\.0){0,2})$

And be done with it.

**Comment by nvdaes** [comment]
Hi Luke, versions maybe added to the nvdaApiVersions file, but not all since add-ons started to be supported in 2012, so previous versions maybe ignored since no add-on can claim to work in NVDA 2011.2, for example.
I think that checking if minimum version exist in API versions prevents someone to set minimumNVDA version, for example, to 2012.5, which doesn't exist.

**Comment by XLTechie** [comment]
But Noelia, making everyone use 0.0.0 for any version earlier than 2019.3.1, 
exactly allows for 2012.5. It does so, by saying that all versions between 0.0.0 
and 2019.3.1 are matched, even if that's not what the author really wants.

Whereas making someone do a PR to modify the API versions file, and have to 
actually figure out whether the version they want constitutes an API change or 
not, or if the change happened in some earlier version, involves a lot of human 
work and time, both on the part of the developer and possibly on the part of NV 
Access.

But, if you really need a hard boundary of 2013.1, you can hardcode a lower 
bound in the check, without otherwise restricting the version to what's in the 
file.

Match against the versions file, and if that match fails, check against the 
regex:

^(201[3-8](\.\d){1,2}|2019\.[12](\.\d)?|0(\.0){1,2})$

should cover 2013.1 through 2019.2.X, or 0.0.0 (or just 0).

It would not allow 2012.5, or 2019.3.1, the latter of course being already 
captured by the exact match with a real API version.

(Note: regex untested, but I'm happy to test and verify if this is the direction 
chosen)

**Comment by nvdaes** [comment]
"Hi Luke: but if we can use just api versions from now, this would mean that 2012.-5 will not be accepted,just 0 dot 0. I supposc tfat, if this is adopted, this will affect just new addons added to the store. I think that the most important think is that nobody sends an add-on with last tested version to 9999. This happened when addonFiles repo was controlled by the community and not by NV Access, and various persons were able to commit there.Enviado desde mi iPhoneEl 5 mar 2023, a las 21:27, Luke Davis ***@***.***> escribió:﻿
But Noelia, making everyone use 0.0.0 for any version earlier than 2019.3.1, 
exactly allows for 2012.5. It does so, by saying that all versions between 0.0.0 
and 2019.3.1 are matched, even if that's not what the author really wants.

Whereas making someone do a PR to modify the API versions file, and have to 
actually figure out whether the version they want constitutes an API change or 
not, or if the change happened in some earlier version, involves a lot of human 
work and time, both on the part of the developer and possibly on the part of NV 
Access.

But, if you really need a hard boundary of 2013.1, you can hardcode a lower 
bound in the check, without otherwise restricting the version to what's in the 
file.

Match against the versions file, and if that match fails, check against the 
regex:

^(201[3-8](\.\d){1,2}|2019\.[12](\.\d)?|0(\.0){1,2})$

should cover 2013.1 through 2019.2.X, or 0.0.0 (or just 0).

It would not allow 2012.5, or 2019.3.1, the latter of course being already 
captured by the exact match with a real API version.

(Note: regex untested, but I'm happy to test and verify if this is the direction 
chosen)

**Comment by XLTechie** [comment]
I'm not talking about last tested version at all.
What I, and also @CyrilleB79 were talking about, was the minimum NVDA version.

I agree with you completely about the last tested version.

**Comment by nvdaes** [comment]
Let"s see what Sean says, since he proposed to use api versions, I think.Enviado desde mi iPhoneEl 5 mar 2023, a las 21:56, Luke Davis ***@***.***> escribió:﻿
I'm not talking about last tested version at all.
What I, and also @CyrilleB79 were talking about, was the minimum NVDA version.

I agree with you completely about the last tested version.

**Comment by seanbudd** [comment]
NVDA does not perform any version checking for add-ons for NVDA versions earlier than 2019.3. This is why the 0.0.0 schema exists.

The add-on store also does not exist in these old versions of NVDA.
I don't understand the value of allowing versions before 2019.3, as:
- these versions are not checked by NVDA for determining add-on compatibility
- these versions don't include the add-on store
- these versions are not supported by NV Access

**Comment by XLTechie** [comment]
@seanbudd wrote:
> The add-on store also does not exist in these old versions of NVDA.

The front-end, no. But will the add-on repository backing the store, also back 
the public download site for add-ons?

Meaning that someone wishing to obtain an add-on for an older version of NVDA, 
will obtain it from the same data source as modern NVDAs obtain their add-ons?

If the store requires to zero out Minimum NVDA Version for versions that 
predate 2019.3, then this public information becomes misleading for add-ons 
which previously only declared themselves to work with, for example, 2018.1 and 
newer.

I think that matters if something other than a store-supporting version of NVDA 
can use the metadata.

My understanding was that the same metadata would indeed be used by other 
processes, such as the public website. If that's wrong, then I probably don't 
have a concern.

**Comment by CyrilleB79** [comment]
> NVDA does not perform any version checking for add-ons for NVDA versions earlier than 2019.3. This is why the 0.0.0 schema exists.


As written by @XLTechie if this data is used by something else than add-on store, e.g. public website, 
if NVDA checks or not add-on compa does not matter since the add-on can be downloaded from elsewhere.


> The add-on store also does not exist in these old versions of NVDA. I don't understand the value of allowing versions before 2019.3, as:

The add-on store does not exist either in 2019.3. Why not begin to 2023.1 or 2023.2 in this case?

> * these versions are not checked by NVDA for determining add-on compatibility

Again, only last version is supported by NVAccess. So lets's add only 2023.1 (or 2022.4 actually for a few days).

More generally, what should be clarified to us is all the situations in which this data need to be used.

**Comment by seanbudd** [comment]
>>these versions are not supported by NV Access
>Again, only last version is supported by NVAccess. So lets's add only 2023.1 (or 2022.4 actually for a few days).

>The add-on store does not exist either in 2019.3. Why not begin to 2023.1 or 2023.2 in this case?

To be clearer, API versions before 2019.3 were never in support by NV Access.
When the API version scheme was created, it was the introduction of a stable add-on API.
Any version of NVDA before that never had an API version.
When manifests indicate this information, it is a guide only to the users who check the manifest, it has no meaningful value as structural data.

Nothing is stopping an author from indicating that sort of information in the add-on description.
However, as metadata values, it is meaningless and misleading for the add-on store infrastructure and NVDA.
The equivalent API version in these cases is 0.0.0 and must be treated as such.

>The front-end, no. But will the add-on repository backing the store, also back
the public download site for add-ons?

NV Access does not have any current plans to host a public website using the add-on data source.
The add-on store is being created with requirements to support NVDA only, however anybody is able to use the data for their own website.

The metadata on API compatibility is not relevant to the legacy add-ons website that NV Access currently hosts.

**Comment by CyrilleB79** [comment]
@seanbudd , sorry, it's still not clear to me.

1. If this data is only used for add-on store and since add-on store will not be present in NVDA before 2023.2, why are there versions suche as 2021.x or 2022.x in the APIs? Said otherwise, with say NVDA 2023.2 or 2023.3, I will have an add-on store; but I will never ask to download an add-on compatible with NVDA 2022.1 through the add-on store.

2. According to change log, the min NVDA version and last tested version info in the manifest have been introduced in NVDA 2019.1, not 2019.3. 2019.3 is just the first version to be tagged as API-breaking, i.e. which has stoppped previous add-ons to work with this new version.

**Comment by seanbudd** [comment]
1. We are still hosting old add-ons that are available on the legacy website.
There is also value in providing some older versions if external contributors wish to host a front-end or backport the add-on store as an add-on.
We also want to hold valid metadata for the add-ons, and having the accurate API version that it was last tested with, even if it is old, is ideal.
We don't want to host non-existent NVDA API versions.

1. Given that information we are going to add 2019.1 and 2019.2 as well.
https://github.com/nvaccess/addon-datastore-transform/pull/19

**Comment by CyrilleB79** [comment]
Many thanks @seanbudd for having taken time to write these details. All is very clear now to me and probably to other people reading this issue.
